lib/puppetserver/ca/action/generate_csr.rb

```diff
@@ -0,0 +1,123 @@
+require 'optparse'
+
+require 'puppetserver/ca/config/puppet'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/local_certificate_authority'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/signing_digest'
+
+module Puppetserver
+  module Ca
+    module Action
+      class GenerateCsr
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = 'Generate a CSR for an intermediate signing CA cert'
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca generate-csr [--help]
+  puppetserver ca generate-csr --output-dir PATH [--config PATH] [--ca-name NAME]
```

Comment:
I think I like the `prepare` name better. Maybe with a `--csr` flag for this behavior? Something like prepare would be too vague to do this w/o specifying an additional flag, I think. But I don't know what would really go with other preparation... creating the directories?? I can only think of "check to make sure my file system's correct" or "check that these artifacts are correct" which are checks we already do in the setup and generate actions.

Or, on the other hand, something more specific like `ca-request`, so `generate-csr` couldn't be construed to mean a regular node csr.

Thoughts @Magisus 	?

Reply:
I like `prepare`, but the fact that we would need something like a `--csr` has a bad smell to me as I think about it. Needing a flag to choose a subaction when we're not sure what else we would use the command for seems confusing. But `ca-request` *also* makes me hesitate because it's not actually a verb, while the rest of the actions are. Something like `generate-ca-request` is almost certainly too long, though, so idk.

Reply:
I think if we're going to have a generic command, we already have one that makes the most sense: `generate`. It already has a special case flag for generating a CA client cert, `--ca-client`, and I don't personally have a problem with giving it a different one, `--ca-csr`, which generates a new key and CSR with CA extensions. The state of the existing CA doesn't matter at this point: it's valid to generate one of these CSRs whether or not you have an existing CA. You're then free to do what you like it with: use it as a new intermediate for your Puppet CA, or use it to bootstrap the Puppet CA with an external root.

I know we've been splitting these actions into "stuff to set up the CA" vs "stuff to do with the CA you already have". But in my mind this one could be either, depending on what you wanted to do with it. In fact, our help text is already misleading about `generate`, because it already has a mode in which the server needs to be stopped, but is still listed in the "require running" [section](https://github.com/puppetlabs/puppetserver-ca-cli/blob/master/lib/puppetserver/ca/cli.rb#L27-L51). So I'm not sure we should tie ourselves to that dichotomy.

It might be useful to get an opinion from CS on this.

Reply:
The other thing that comes up with using `generate` is the question of what the behavior should be. Right now with the separate command, it writes out to a required specified directory, but that doesn't seem appropriate if we use `generate`­­­—conditionally required parameters seem like a bad idea. I'm not sure what would be appropriate behavior in that case.